proposals/0000-linear-types.rst

```diff
@@ -0,0 +1,858 @@
+   x ⩽ y = x == y
+
+Every variable in the environment is annotated with its multiplicity,
+which constrains how it can be used. A variable usage is said to be
+multiplicity ``p`` in a term ``u`` if:
+
+- ``p=0`` and ``x`` is not free in ``u``
+- ``p=1`` and ``u = x``
+- ``p=p1+q*p2`` `` and ``u = u1 u2`` with ``u1 :: a :q-> b`` and the
+  usage of ``x`` in ``u1`` is ``p1``, and in ``u2`` is ``p2``
+- ``u = λy. v`` and the usage of ``x`` in ``v`` is ``p``.
+
+A variable's usage is correct if it is smaller than or equal to the
+multiplicity annotation of the variable. Incorrect usage results in a
+type error.
+
+The multiplicity of a variable introduced by a λ-abstraction is taken
+from the surrounding typing information (typically a type annotation
+on an equation). For instance
+
+::
+
+  foo :: A :p-> B
+  foo x = …  -- x has multiplicity p
+
+The above takes care of the pure λ-calculus part of Haskell. We also
+need to consider ``let`` and ``case``.
+
+A ``let`` binding is considered to have an implicit multiplicity
+annotation (the annotation is inferred). The variables introduced by a
+``let`` bindings with annotation ``p`` all have multiplicity
+``p``. And the usage of ``x`` in ``let_p {y1 = u1; … ;yn = un} in v``
+(where the ``yi`` are variables) is ``p*q1 + … + p*qn + q`` where the
+usage of ``x`` in ``ui`` is ``qi`` and in ``v`` is ``q``.
+
+If a let has recursive binders, then ``p`` must be ``ω``.
+
+A ``case`` expression has an implicity multiplicity annotation, like
+``let`` binding. It if often inferred from the type annotation of an
+equation. The usage of ``x`` in ``case_p u of { … }`` where the usage
+of ``x`` in ``u`` is ``q`` is ``p*q`` plus the *join* of the usage of
+``x`` in each branch.
+
+The multiplicity annotation of variables introduced by a pattern depend
+on the constructor and on the implicit annotation of the
+``case``. Specifically in ``case_p u of {…; C x1 … xn -> …; …}`` Where ``C :: a1 :q1-> … an :qn-> A``,
+Then ``xi`` has multiplicity annotation ``p*qi``. For instance
+
+::
+
+  bar :: (a,b) :p-> c
+  bar (x,y) = … -- Since (,) :: a ->. b ->. (a,b), x and y have
+                -- multiplicity p
+
+The type ``A->.B`` is a strengthening of ``A->B``, but the type
+checker doesn't do subtyping. It relies on polymorphism
+instead. However, following the definition above, note that
+
+::
+
+  f :: A ->. B
+
+  g :: A -> B
+  g = f  -- should not be well-typed
+  g x = f x  -- is well-typed
+
+This would be rather unfortunate: for instance a linear function in a
+library could not be used with ``map`` from base. Which means that
+everybody would have to start caring about linearity. Worse: every use
+of ``map Just`` would now be untyped. Fortunately, this sort of
+opportunity is easily detected and the former definition of ``g`` is
+understood as the latter, well-typed, one. It means that is not a
+breaking change to strengthen a *first-order* regular arrow ``->``
+into a linear ``->.`` in an interface.
+
```

Comment:
(This relaxation allows a smooth migration path from regular libraries to linearly-typed ones)